crates/c-api/include/wasmtime/async.h

```diff
@@ -0,0 +1,265 @@
+ * continue to yield WebAssembly execution.
+ */
+typedef bool (*wasmtime_func_async_continuation_callback_t)(void *env);
+
+/**
+ * A continuation for the current state of the host function's execution.
+ */
+typedef struct wasmtime_async_continuation_t {
+  /// Callback for if the async function has completed.
+  wasmtime_func_async_continuation_callback_t callback;
+  /// User-provided argument to pass to the callback.
+  void *env;
+  /// A finalizer for the user-provided *env
+  void (*finalizer)(void *);
+} wasmtime_async_continuation_t;
+
+/**
+ * \brief Callback signature for #wasmtime_linker_define_async_func.
+ *
+ * This is a host function that returns a continuation to be called later. 
+ * The continuation returned is owned by wasmtime and will be deleted when it completes.
+ *
+ * All the arguments to this function will be kept alive until the continuation
+ * returns that it has errored or has completed.
+ *
+ * \param env user-provided argument passed to #wasmtime_linker_define_async_func
+ * \param caller a temporary object that can only be used during this function
+ * call. Used to acquire #wasmtime_context_t or caller's state
+ * \param args the arguments provided to this function invocation
+ * \param nargs how many arguments are provided
+ * \param results where to write the results of this function
+ * \param nresults how many results must be produced
+ * \param trap_ret if assigned a not `NULL` value then the called function will
+ *        trap with the returned error. Note that ownership of trap is transferred
+ *        to wasmtime.
+ *
+ * Only supported for async stores.
+ *
+ * See #wasmtime_func_callback_t for more information.
+ */
+typedef wasmtime_async_continuation_t *(*wasmtime_func_async_callback_t)(
+    void *env,
+    wasmtime_caller_t *caller, 
+    const wasmtime_val_t *args,
+    size_t nargs, 
+    wasmtime_val_t *results,
+    size_t nresults,
+    wasm_trap_t** trap_ret);
+
+/**
+ * \brief The structure representing a asynchronously running function.
+ *
+ * This structure is always owned by the caller and must be deleted using #wasmtime_call_future_delete.
+ *
+ * Functions that return this type require that the parameters to the function are unmodified until 
+ * this future is destroyed.
+ */
+typedef struct wasmtime_call_future wasmtime_call_future_t;
+
+/**
+ * \brief Executes WebAssembly in the function.
+ *
+ * Returns true if the function call has completed. After this function returns true, it should *not* be 
+ * called again for a given future.
+ *
+ * This function returns false if execution has yielded either due to being out of fuel 
+ * (see wasmtime_store_out_of_fuel_async_yield), or the epoch has been incremented enough 
+ * (see wasmtime_store_epoch_deadline_async_yield_and_update).
+ *
+ * The function may also return false if asynchronous host functions have been called, which then calling this 
+ * function will call the continuation from the async host function.
```

Comment:
Mind combining this clause with the above one so the three cases of suspension are enumerated as a set?

Reply:
Done.